### DISCUSSION TITLE
Allow setting COOP and COEP headers in Github Pages

### DISCUSSION THREAD

**Discussion by josephrocca** [community/community#13309] [Other Feature Feedback, Questions, & Ideas]
In order to use `SharedArrayBuffer` and WebAssembly Threads, COOP and COEP headers need to be set, such that the page becomes cross-origin isolated ([explanation at web.dev](https://web.dev/coop-coep/)). Since Wasm is becoming well supported across browsers, many libraries (e.g. [image editing](https://github.com/kleisauke/wasm-vips), [ML](https://github.com/microsoft/onnxruntime/tree/master/js/web)) and applications are starting to use it, and so the inability to set these headers in Github Pages is becoming a bigger and bigger problem (for me, at least). 

Note that there are hacky workarounds, like this: https://github.com/gzuidhof/coi-serviceworker But these can cause a bunch of other problems. I think the emergence of libraries like that (which explicitly mention Github Pages in their README) is evidence enough that this feature should be added.

Thanks!

**Comment by mithrendal** [comment]
I want this too ...

**Comment by josephrocca** [comment]
For people who want a quick and **very** hacky workaround, include this script in your `<head>`:

https://github.com/josephrocca/clip-image-sorter/blob/main/enable-threads.js

(Note: You can't paste it inline - it needs to be in a separate script file and imported like `<script src="enable-threads.js"></script>`)

**Reply by utkrssh** [reply]
This works in Chrome, but not in Firefox. Any idea why?

**Reply by utkrssh** [reply]
Setting `require-corp` instead of `credentialless` as `Cross-Origin-Embedder-Policy` makes it works in both, Chrome and Firefox.

**Reply by josephrocca** [reply]
Yep, `credentialless` is the best value to use here (because it doesn't block loading images from a cross-domain CDN, and other stuff like that), but it's quite new and not supported on Firefox or Safari yet, as of writing: https://caniuse.com/mdn-http_headers_cross-origin-embedder-policy_credentialless

**Reply by iamshreeram** [reply]
@josephrocca , 

Thank you for sharing the script, this will be helpful if it works. I'm testing it on this page - https://shreeram.us/video.transcoder/. Unfortunately, it doesn't seem to be working for my specific situation in both Chrome and Firefox. 

In Chrome, the "else" part is being triggered frequently, causing the page to continually reload. 
In Firefox, I'm getting the following error message: Uncaught (in promise) TypeError: navigator.serviceWorker is undefined at line 53 and 68 of enable-threads.js. 

I am using `credentialless` as it is supported in both Chrome and Firefox. 

Appreciate any help here. thanks!

**Reply by vitaly-castLabs** [reply]
Doesn't work for me. It gets me around SharedArrayBuffer checks, but nothing actually works

**Reply by stereomatchingkiss** [reply]
> Doesn't work for me. It gets me around SharedArrayBuffer checks, but nothing actually works

Use netlify, remember to place header [like this one](https://github.com/stereomatchingkiss/fall_down_detection/blob/main/_headers)

**Reply by chn-lee-yumi** [reply]
It works for me! Many thanks!

**Comment by mithrendal** [comment]
@josephrocca my app is already using an own service worker which would conflict with the one you proposed... I also already tricked it to overwrite the headers ... nonetheless this feels still very hacky ... would prefer a straight forward solution where github would offer to allow to enable us COOP and COEP headers. Your script is interesting, mine for example did not consider `request.mode == 'no-cors'` ... thank you very much 👍, will continue to study and learn from it

**Comment by coatless** [comment]
Any thoughts as to whether something akin to [Netlify's header specification](https://docs.netlify.com/routing/headers/#syntax-for-the-netlify-configuration-file) could be added? The lack of being able to address   Cross-Origin-Opener-Policy (COOP) and  Cross-Origin-Embedder-Policy (COEP) on a pro-developer portal is surprising.

**Comment by yoannchaudet** [comment]
Hello,

This is a scenario we would support with custom headers. No ETA at the moment as to when this feature will be available.

**Reply by coatless** [reply]
Wonderful! Could we expect an approach like [Netlify's header specification `_headers`](https://docs.netlify.com/routing/headers/#syntax-for-the-netlify-configuration-file) to specify the custom header? Also, is there a way to look at a roadmap or public timeline view for updates? Or will you be updating the progress in this thread?

**Reply by Cundiff7** [reply]
Hi :)
Has there been any update to the timing of this feature? Thanks!

**Reply by HarikrishnanBalagopal** [reply]
Same question here. Trying to use `SharedArrayBuffer` with `Atomics.wait` for async stuff in WebAssembly.
However since Github Pages isn't `crossOriginIsolated` it's not available.

Any update on if we can get at least a CORS headers checkbox in the Settings > GH Pages tab of a repo?

**Reply by webJose** [reply]
Hello!  As all others, I'm in the same boat.  I created [@wjfe/async-workers](https://github.com/WJSoftware/wjfe-async-workers), and contains 5 synchronization objects based on `Atomics` and `SharedArrayBuffer` that I would like to showcase in its demo deployed in GitHub Pages, but currently cannot.

Is there an open issue that we can track ourselves regarding the feature that allows adding HTTP headers?

**Reply by stereomatchingkiss** [reply]
Use netlify, remember to place header [like this one](https://github.com/stereomatchingkiss/fall_down_detection/blob/main/_headers)

**Comment by stereomatchingkiss** [comment]
> Hi :) Has there been any update to the timing of this feature? Thanks!

Use netlify

**Reply by Cundiff7** [reply]
I ported to netlify, but i am still having trouble. My goal is to embed an HTML 5 game onto a webpage via itch.io. I have tried a few different things for the headers, but have not had luck. Any tips?

**Reply by utkrssh** [reply]
```
[[headers]]
for = "/*"
  [headers.values]
    Cross-Origin-Opener-Policy = "same-origin"
    Cross-Origin-Embedder-Policy = "require-corp"
```
This had worked for me.

**Reply by Cundiff7** [reply]
I will check it out, thanks!

**Comment by Gurupatil0003** [comment]
It's Osm

**Comment by joyel24** [comment]
Would be nice for Godot HTML5 export !

**Reply by westurner** [reply]
- From https://github.com/ray-pH/quantumQ/issues/5 :

> - https://docs.godotengine.org/en/stable/tutorials/export/exporting_for_web.html#serving-the-files
>   * https://github.com/gzuidhof/coi-serviceworker :
>     > Cross-origin isolation (COOP and COEP) through a service worker for situations in which you can't control the headers (e.g. GH pages)

**Comment by DesignThinkerer** [comment]
Any update on this ? I'm attempting to setup a github page with wasm+sqlite, I'll share here if the workarounds mentioned in this thread are still working or not.

**Comment by moalhaddar** [comment]
Any updates? I want to publish a wasm example from my repository but i'm limited by this.

**Reply by joyel24** [reply]
> Any updates? I want to publish a wasm example from my repository but i'm limited by this.

Maybe if we ping @microsoft @github all together ! 😁

**Comment by gregdan3** [comment]
Working on a data vis project that would greatly benefit from this! Mostly in response times, ehe.

**Comment by floooh** [comment]
I would also greatly appreciate that for my WASM demos hosted on Github pages (e.g. https://floooh.github.io/sokol-html5/ or https://floooh.github.io/visualz80remix/), IMHO not being able to control COOP/COEP headers on simple web hosting services like Github is the one thing that's holding back WASM threading.

**Comment by ayan4m1** [comment]
The hacky workaround no longer works, so please can we have the ability to do this?

**Comment by floooh** [comment]
@ayan4m1 out of interest, which workaround has stopped working? The service-worker thing? If yes that would be really bad....

**Reply by ayan4m1** [reply]
<img width="940" height="385" alt="image" src="https://github.com/user-attachments/assets/f0a27f72-4842-4a07-b8ae-cebf177d05e1" />

Using `coi-serviceworker v0.1.7` in latest Firefox for Windows

EDIT: Might be because I was serving it inline, but trying to bundle this script is a nightmare, so I still haven't found a way to make it work.

**Comment by alshdavid** [comment]
+1 please.

My current "workaround" is to deploy an AWS S3 website + CloudFront distribution and set the headers there. Which isn't a workaround, it's just not using GitHub Pages at all.

**Reply by webJose** [reply]
Hehe, this is by far the most stretching the definition of "workaround" has ever seen. 😄  Thanks for the chuckles.  Cheers! 👍

**Comment by Crisspl** [comment]
+1, i really need it as well

**Comment by samuelpmish** [comment]
Also affected-- annoying to lose out on important WASM features with gh pages sites!
